### PR TITLE
Restrict visibility of prometheus packages

### PR DESCRIPTION
**Pull request by serathius**
/kind feature

**What this PR does / why we need it**:
This PR restricts visibility of prometheus packages which no longer should be directly imported.

**Which issue(s) this PR fixes**:
Refer kubernetes/enhancements#1238

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
/milestone v1.17
/priority important-soon
/assign RainbowMango

### CODE REVIEW

File: vendor/BUILD [hunk 1]
```diff
@@ -319,12 +319,7 @@ filegroup(
         "//vendor/github.com/pkg/errors:all-srcs",
         "//vendor/github.com/pmezard/go-difflib/difflib:all-srcs",
         "//vendor/github.com/pquerna/cachecontrol:all-srcs",
-        "//vendor/github.com/prometheus/client_golang/prometheus:all-srcs",
-        "//vendor/github.com/prometheus/client_model/go:all-srcs",
-        "//vendor/github.com/prometheus/common/expfmt:all-srcs",
-        "//vendor/github.com/prometheus/common/internal/bitbucket.org/ww/goautoneg:all-srcs",
-        "//vendor/github.com/prometheus/common/model:all-srcs",
-        "//vendor/github.com/prometheus/procfs:all-srcs",
+        "//vendor/github.com/prometheus:all-srcs",
```
Comment (liggitt) [inline vendor/BUILD]: was this hand-edited? won't the hack/update-bazel script undo these changes?

**Comment by RainbowMango** [comment]
I don't know too much about bazel rule. 
I'm not quite sure if we can modify files under `vendor/`.

And I think we should wait for all migration done.

File: vendor/BUILD (hunk 1 above)
Comment (serathius) [reply vendor/BUILD]: After I added BUILD file with visibility rules to `vendor/github.com/prometheus` hack/update-bazel script added `all-srcs` rules to it and replaced it `vendor/BUILD`

**Comment by munnerz** [comment]
This seems like a fairly reasonable approach, although it does mean that if a user runs `rm -Rf vendor/` and `hack/update-vendor.sh`, their vendor directory will be incorrect/not up to date with what it was before.

Alternatively, could you define:

```
package_group(
    name = "allow_list",
    includes = [":internal"],
    packages = [
        "//cluster/images/etcd-version-monitor",
        "//pkg/controller/endpointslice",
        "//pkg/controller/volume/attachdetach/metrics",
        "//pkg/controller/volume/persistentvolume/metrics",
        "//pkg/kubelet/metrics/collectors",
        "//pkg/kubelet/pleg",
        "//pkg/kubelet/pluginmanager/metrics",
        "//pkg/kubelet/server",
        "//pkg/kubelet/server/stats",
        "//pkg/kubelet/volumemanager/metrics",
        "//pkg/master",
        "//pkg/scheduler/framework/v1alpha1",
        "//pkg/volume/util/operationexecutor",
        "//staging/src/k8s.io/apiserver/pkg/admission/metrics",
        "//staging/src/k8s.io/component-base/metrics/...",
        "//test/e2e",
        "//test/e2e/storage",
        "//test/e2e_node",
        "//vendor/...",
    ],
)
```

in a package *not* in `vendor/`, and then use `buildozer` to update the `visibility` setting in all packages under `vendor/github.com/prometheus` to be the required value? You would likely want to add the additional call to buildozer here: https://github.com/kubernetes/kubernetes/blob/master/hack/update-bazel.sh#L82

This would then mean that `git clone k/k.git && rm -Rf vendor/ && ./hack/update-vendor.sh` would result in a clean tree.

**Comment by serathius** [comment]
@munnerz This is a great suggestion! Thanks!

File: vendor/BUILD (hunk 1 above)
Comment (serathius) [reply vendor/BUILD]: Thanks to @munnerz suggestion I updated `hack/update-bazel.sh` to set visibility.

**Comment by munnerz** [comment]
From a Bazel and general repo management, this looks good to me 😄 with regards to whether or not sig-instrumentation want to do this/whether this should be *approved* is not up to me 😄 

So adding my lgtm and leaving to the appropriate OWNERS to decide on approval...

/lgtm

**Comment by BenTheElder** [comment]
/approve
for hack/

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve
